src/Lean/Server/Watchdog.lean

```diff
@@ -307,6 +307,9 @@ section ServerM
   def terminateFileWorker (uri : DocumentUri) : ServerM Unit := do
     let fw ← findFileWorker! uri
     try
+      let ctx ← read
+      let meta := fw.doc.meta
+      publishDiagnostics meta #[] ctx.hOut
```

Comment:
I think this belongs in `forwardMessages`, where we know the file worker has actually exited and won't send out diagnostics of its own anymore. Though as @gebner mentioned afair, we still have other races such as not waiting for the old file worker to exit before starting the new one.

Reply:
@Vtec234 Thoughts? Maybe it's okay to leave it like this for now, at least we won't race with diagnostics from the new file worker process then...

Reply:
Hm I have unfortunately lost any mental model of what order the watchdog does things in a while ago, but I think you are right, we should wait for the worker to exit. I am not sure if we should also clear the diagnostics when the worker crashes. Maybe only doing it on termination is better, since when it crashes the "leftover" diagnostics could be helpful in diagnosing the issue.

Reply:
*Ideally*, I think the diagnostics should only be cleared if the file is closed.  If the worker crashes, we should preserve the current diagnostics, show in the progress bar which part of the file was not processed due to crashing, and maybe even add an error message about the crash to the diagnostics.

The race conditions require a larger refactoring anyhow, so I'm not going to think about the best location for the bandaid.

Reply:
Although this particular one here is an even larger race condition than usual...

Reply:
Are you saying where it was before in FileWorker was better because that way we can ensure it is the last thing the `mainLoop` does before exiting?

Reply:
The corresponding watchdog location should be before https://github.com/leanprover/lean4/blob/9f29967fb0464542cf524bbef56ae598db71fc54/src/Lean/Server/Watchdog.lean#L259

Reply:
> Although this particular one here is an even larger race condition than usual...

It would be relatively simple to fix though - we'd have to move the `startFileWorker` call from `handleEdit` to the location I just mentioned above

Reply:
Ok, folks, I did some work, along the lines outlined by Sebastian above, to ensure publish diagnostics is always the last thing sent when a document is closed, by adding a new WorkerState.Exiting (changing state to an IO.Ref) so forwardMessages can see this state change and ensure that the very last thing we do is clear the diagnostics).  

I tested it with some large Lean files that take a while to parse, quickly closing them before all the messages have been sent, and I could not get any left-over messages to remain in the output window - trying different timing on when I closed the documents.  I also tested opening multiple lean source files and closing them all at once.  Do you guys have any suggestions on what would be the trickiest test case we can try for this?

Reply:
never mind, I moved the clear diagnostics to [lean4/src/Lean/Server/Watchdog.lean](https://github.com/leanprover/lean4/blob/9f29967fb0464542cf524bbef56ae598db71fc54/src/Lean/Server/Watchdog.lean#L259)